### ISSUE TITLE
Query about the new update about inference on CPU and GPU.

### ISSUE THREAD

**Issue by kaivu1999** [hpi-xnor/BMXNet-v2#5]
## Description
I am actually interested in the speed up that I can get on CPU and GPU especially for inference.

According to the [answer](https://github.com/hpi-xnor/BMXNet/issues/6#issuecomment-305904256) by @yanghaojin 
I have tried BMXNet v1 for the same and I get speed up on CPU of about 1.4x - 1.7x on my PC for some models but also a decrease in speed up in some case.
I used : Ubuntu 16.04/64-bit platform on Intel(R) Core™ i5-8250U CPU@ 1.60GHz (supports SSE4.2)

## Can you please elaborate about the update of 21st May 2019 wrt speed up ?
The update which is written in the[ changelog ](https://github.com/hpi-xnor/BMXNet-v2/blob/master/CHANGELOG.md#030---2019-05-21)

**Comment by simonmaurer** [comment]
@kaivu1999 that's a really good discussion you brought up and has been of interest since the release of BMXNetv1.
pointing out the findings of @yanghaojin that for training you skipped the GEMM computation and use the CuDNN implementation for training on GPUs (as per your justification), theoretically for inference (say real-time constraints on robotic platforms) XNOR computations could be speeded up on GPUs as well - did you find the reasons for this ? (eg. well optimized CuDNN code by teams of nVidia?, etc.)
now only considering CPU inference which is the main focus of the whole research (as the ultimate goal is to run DL models preferably in realtime on CPU-only devices):
the results of @kaivu1999 dont coincide with the results of the BMXNet papers and preceeding XNORNet paper (theoretically up to 32x)
clearly there are other steps involved like input binarization or patch2col that add up to the total processing time as part of the convolutional operation
@yanghaojin, @jopyth would be interested as well if you could elaborate a bit on that?

**Comment by kaivu1999** [comment]
@simonmaurer I think in the XNORNet they mention 32x memory savings and 52x computation speeding.
I would also like to ask specific which specific files should I look for in the repository to understand the optimization part.

**Comment by yanghaojin** [comment]
> @kaivu1999 that's a really good discussion you brought up and has been of interest since the release of BMXNetv1.
> pointing out the findings of @yanghaojin that for training you skipped the GEMM computation and use the CuDNN implementation for training on GPUs (as per your justification), theoretically for inference (say real-time constraints on robotic platforms) XNOR computations could be speeded up on GPUs as well - did you find the reasons for this ? (eg. well optimized CuDNN code by teams of nVidia?, etc.)
> now only considering CPU inference which is the main focus of the whole research (as the ultimate goal is to run DL models preferably in realtime on CPU-only devices):
> the results of @kaivu1999 dont coincide with the results of the BMXNet papers and preceeding XNORNet paper (theoretically up to 32x), maybe we can verify this with BMXNetv2..
> clearly there are other steps involved like input binarization or patch2col that add up to the total processing time as part of the convolutional operation
> @yanghaojin, @Jopyth would be interested as well if you could elaborate a bit on that?

using CuDNN for training because it is over optimized and much faster than the preliminary implementation of the xnor kernel in cuda, I believe an optimized xnor cuda kernel will indeed improve the speed a lot, but we are not expert for that thus leave this to the community for now. 
In our first paper, the evaluation we mainly compared the xnor-gemm function with the traditional DOT operators using CPU. so xnor cpu kernel is indeed much faster than the standard dot engine like CBlas, but as I mentioned [here](https://github.com/hpi-xnor/BMXNet/issues/6) there is also huge optimization potential regarding the whole convolution layer. In BMXNetV1 our evaluation showed that the portion of xnor-gemm computation is too small compared to the overall computations of a single Qconv layer (see this [screenshot](https://www.dropbox.com/s/roedt86tn69ucnx/Bildschirmfoto%202019-06-13%20um%205.19.03%20PM.png?dl=0) this result might be outdated but I think the ratio of each part doesn't change much).

According to the 52 times speedup mentioned in the xnor paper. I think this number is also based on the xnor-gemm function, not the whole conv layer. And they only compared to the naive implementation of a dot engine, even without reporting the comparison with CBlas (atlas or others). I have seen their code in Darknet before they removed them years ago (since they initialized a startup company XNOR.ai thus remove the code from darknet).

**Comment by kaivu1999** [comment]
Thank you very much,
@yanghaojin  for a detailed explanation.

So I would like to ask a few things.
1. In BMXNet paper, the code example for LeNet and Binary LeNet where a specific order for layers is used (QActivation-QConv/QFC-BatchNorm-Pooling). Is this has to play any role in optimization ?
2. So is input_binarisation done once and all the layers follow or Do we have to binarize the input layers every time for each layer? 
3. There are these other layers like Pooling and BatchNorm which don't have their accelerated counterparts as I see only (QActivation,QConvolution,QFullyconnected). So do they cause unnecessary data type conversion (if we expect that binarisation happens only once for all consecutive activation layers, which is the whole point right ?)
4. Is there an effect of batch_size for inference ( Theoretically it should not be) but I get an increase in speedup with an increase in batch_size?

Also I got these numbers for BMXNet v1 considering only inference.

NIN-Cifar10          |  VGG-11-Cifar10
:-------------------------:|:-------------------------:
![NIN](https://user-images.githubusercontent.com/35027192/59478848-63f45280-8e8d-11e9-94f9-5a7606f5bdb5.png)  | ![VGG-11](https://user-images.githubusercontent.com/35027192/59478849-65be1600-8e8d-11e9-9fa6-619512aa98dc.png)

Where the blue line represents the experiment where I tried to use the accelerated layers as much as possible. I accelerated the last layer (Not recommended for accuracy) and for NIN I had to approximate some of the layers in between so that the input channels are a multiple of 64 as the accelerated seems to support input layers with size in multiples of 64.

Also for a Network with just Convolution and Activation layer pairs. In a network with 8 of such pairs where I accelerated 7 of them in the binary version, I am getting a Speed Up of 6.8x for 256 as batch_size. 

Also I would really like to know about the update on similar notes ?

> I would also like to ask specific which specific files should I look for in the repository to understand the optimization part?
